README.md
# LoggerPapertrailBackend

[![Hex.pm](https://img.shields.io/hexpm/v/logger_papertrail_backend.svg?maxAge=2592000)](https://hex.pm/packages/logger_papertrail_backend)
[![Hex.pm](https://img.shields.io/hexpm/dt/logger_papertrail_backend.svg)](https://hex.pm/packages/logger_papertrail_backend)
[![Build Status](https://travis-ci.org/larskrantz/logger_papertrail_backend.svg?branch=master)](https://travis-ci.org/larskrantz/logger_papertrail_backend)

A [Papertrail](https://papertrailapp.com) backend for [Elixir Logger](http://elixir-lang.org/docs/stable/logger/Logger.html).

[Changelog](CHANGELOG.md)

## Installation

Available in [Hex](https://hex.pm/packages/logger_papertrail_backend). The package can be installed as:

* Add `logger_papertrail_backend` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [{:logger_papertrail_backend, "~> 1.1"}]
end
```

* Ensure `logger` and `logger_papertrail_backend` is started before your application (pre Elixir 1.4):

```elixir
def application do
  [applications: [:logger, :logger_papertrail_backend]]
end
```

* Or after Elixir 1.4, just ensure `Logger` is in `extra_applications`:

```elixir
def application do
  [extra_applications: [:logger]]
end
```

* In your `config.exs` (or in your `#{Mix.env}.exs`-files):

```elixir
config :logger, :logger_papertrail_backend,
  host: "logs.papertrailapp.com:<port>",
  level: :warn,
  system_name: "Wizard",
  metadata_filter: [],
  format: "$metadata $message"
```

  Alternatively use :url for shorter config.
  Prepend with "papertrail://" (or whatever as long its a valid URI), then host:port/system_name. We normally set an ENV-var: `url: System.get_env("PAPERTRAIL_URL")` or `url: "${PAPERTRAIL_URL}",` if building releases with [Distillery](https://github.com/bitwalker/distillery) and `REPLACE_OS_VARS=true`.

```elixir
config :logger, :logger_papertrail_backend,
  url: "papertrail://logs.papertrailapp.com:<port>/<system_name>",
  level: :warn,
  format: "$metadata $message"
```

  Then config `:logger` to use the `LoggerPapertrailBackend.Logger`:

```elixir
config :logger,
  backends: [ :console,
    LoggerPapertrailBackend.Logger
  ],
  level: :debug
```

  _Note: if you have an umbrella project, use your top `config.exs`._

* (Required) Follow "Add System" in your Papertrail dashboard to get `:host` values
* (Optional) Set `:level` for this backend (overides global `:logger`-setting )
* (Optional) Set specific `:system_name` in Papertrail, defaults to current application-name
* (Optional) Set :format, defaults to `[$level] $levelpad$metadata $message`, see [Logger.Formatter](https://hexdocs.pm/logger/Logger.Formatter.html#content)
* (Optional) Set `:metadata_filter` - metadata terms which must be present in order to log. See [LoggerFileBackend](https://github.com/onkel-dirtus/logger_file_backend#filtering-specific-metadata-terms) for examples.
* Other supported options in config are `:colors`, `:metadata`. See :console-docs in [Logger](https://hexdocs.pm/logger/Logger.html#module-console-backend)

## Example output

`Oct 02 14:19:04 Wizard UpptecSlack.SlackBot:  [info]   Successfully authenticated as user "wizard" on team "Upptec"`

Papertrail sets timestamp when message arrives. `Wizard` is `:system_name`. `UpptecSlack.SlackBot` is the module sending the log. `[Info]` is level.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.1.0] - 2019-03-27

### Changed

- Merged #19 from @ivanovv and @lasseebert, adding `metadata_filter` option.

## [1.0.3] - 2018-03-06

### Changed

- Merged #16 from @mkrogemann, squashing `warning: outdented heredoc line` in Elixir 1.6

## [1.0.2] - 2017-12-12

### Fixed

- Issue #15, map elixir level :warn to correct syslog level

## [1.0.1] - 2017-10-09

### Fixed

- Issue #14, crashes when receiving unhandled messages in handle_info. Now ignoring unhandled `info`, `events`, `code_change` and `terminate` to actually behave as a `:gen_event`.

## [1.0.0] - 2017-10-05

### Changed

- Using behavior `:gen_event` since `GenEvent` is deprecated
- Squash warnings from deprecated String-functions
- Requires Elixir ~> 1.3

## [0.2.1] - 2017-03-19

### Fixed

- An issue with configuration, issue #9

## [0.2.0] - 2017-03-17

### Changed

- Sender will now hold a port open in GenServer instead of opening and closing per request.
- More informative messages if trying to configure with faulty config.

## [0.1.2] - 2017-03-06

### Changed

- Merged #7 squashing warnings from Elixir 1.4.
- Some module documentation

## [0.1.1] - 2016-10-24

### Changed

- Fixed issue #4, if DNS resolution fails (ip is resolved every 60s), it uses last successfully resolved ip.

## [0.1.0] - 2016-04-18

### Added

- Change Log
- Possibility to use `:url` to config log target. Format: `papertrail://<host>:<port>/<system_name>` instead of `:host` and `:system_name`

### Changed

- Always trims away leading `Elixir.` in module name (tag in [Papertrail](http://papertrailapp.com) semantic)
- Since [Papertrail](http://papertrailapp.com) tags has a limitation of 32 chars, `LoggerPapertrailBackend` will trim away module names from the start of the name, ie `This.Is.Ridiculously.Long.Module.Name` becomes `Is.Ridiculously.Long.Module.Name` in the logs.

## [0.0.1 / 0.0.2] - 2015-10-02

First stable releases.


lib/logger_papertrail_backend/message_builder.ex
defmodule LoggerPapertrailBackend.MessageBuilder do
  @moduledoc """
    Handles building messages that will be sent to Papertrail from Elixir
  """
  @doc """
  Will build a syslog-message, roughly conforming to a BSD syslogmessage, RFC 3164.
  But it is specially fitted to work with PaperTrail, http://papertrailapp.com

  ## Example

      iex> LoggerPapertrailBackend.MessageBuilder.build("Hello PaperTrail!", :error, "my_system_name", {{2015,10,1}, {12,44,1}}, "Elixir.Hello.World")
      "<11>Oct  1 12:44:01 my_system_name Hello.World Hello PaperTrail!"
  """

  @spec build(message :: binary, level :: atom, hostname :: binary, timestamp :: tuple, tag :: any) :: binary
  def build(message, level, hostname, timestamp, tag) do
    facility = :user # Papertrail does not care it seems, so just use :user
    priority = calculate_priority(facility, level)
    bsd_timestamp = create_bsd_timestamp(timestamp)
    trimmed_tag = trim_tag(tag)
    "<#{priority}>#{bsd_timestamp} #{hostname} #{trimmed_tag} #{message}"
  end

  defp create_bsd_timestamp({date,{h,m,s,_ms}}), do: create_bsd_timestamp({date,{h,m,s}})
  defp create_bsd_timestamp({{_y,mo,d},{h,m,s}}) do
    months = ~W{Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec}
    padded_day = d |> Integer.to_string |> String.pad_leading(2)
    month = months |> Enum.at(mo-1)
    zeropad = fn(i) -> i |> Integer.to_string |> String.pad_leading(2, "0") end
    "#{month} #{padded_day} #{zeropad.(h)}:#{zeropad.(m)}:#{zeropad.(s)}"
  end

  defp trim_tag(tag) when is_binary(tag) do
    stripped = tag |> String.trim() |> String.replace("Elixir.", "")
    trim_tag(stripped, String.length(stripped))
  end
  defp trim_tag(tag), do: trim_tag("#{tag}")

  defp trim_tag(tag, current_length) when current_length <= 32, do: tag
  defp trim_tag(tag, _) do
    modules = tag |> String.split(".")
    new_tag = trim_module_name(modules) |> Enum.join(".")
    trim_tag(new_tag, String.length(new_tag))
  end

  defp trim_module_name(modules), do: trim_module_name(modules, [])
  defp trim_module_name([], state), do: Enum.reverse(state)
  defp trim_module_name([ last_single_module | []], []), do: [ String.slice(last_single_module, 0, 32) ]
  defp trim_module_name([ _head | tail ], state), do: (state |> Enum.reverse) ++ tail


  # https://tools.ietf.org/html/rfc3164#section-4.1.1

  defp calculate_priority(facility, level), do: facility(facility) * 8 + level(level)

  # Thanks to https://github.com/jkvor/erlang_syslog/blob/master/src/syslog.erl

  defp level(:warn), do: level(:warning) # map elixir :warn to :warning
  defp level(:emergency), do: 0 # system is unusable
  defp level(:alert), do: 1 # action must be taken immediately
  defp level(:critical), do: 2 # critical conditions
  defp level(:error), do: 3 # error conditions
  defp level(:warning), do: 4 # warning conditions
  defp level(:notice), do: 5 # normal but significant condition
  defp level(:info), do: 6 # informational
  defp level(:debug), do: 7 # debug-level messages
  defp level(_), do: level(:info) # default to info

  # % paraphrased from https://github.com/ngerakines/syslognif/blob/master/src/syslog.erl#L55
  defp facility(:kern), do: 0      # kernel messages
  defp facility(:user), do: 1      # random user-level messages
  defp facility(:mail), do: 2      # mail system
  defp facility(:daemon), do: 3    # system daemons
  defp facility(:auth), do: 4      # security/authorization messages
  defp facility(:syslog), do: 5    # messages generated internally by syslogd
  defp facility(:lpr), do: 6       # line printer subsystem
  defp facility(:news), do: 7      # network news subsystem
  defp facility(:uucp), do: 8      # UUCP subsystem
  defp facility(:cron), do: 9      # clock daemon
  defp facility(:authpriv), do: 10 # security/authorization messages (private)
  defp facility(:ftp), do: 11      # ftp daemon

  defp facility(:local0), do: 16   # reserved for local use
  defp facility(:local1), do: 17   # reserved for local use
  defp facility(:local2), do: 18   # reserved for local use
  defp facility(:local3), do: 19   # reserved for local use
  defp facility(:local4), do: 20   # reserved for local use
  defp facility(:local5), do: 21   # reserved for local use
  defp facility(:local6), do: 22   # reserved for local use
  defp facility(:local7), do: 23   # reserved for local use
  defp facility(_), do: facility(:user)
end
